deque.go

```diff
@@ -242,6 +242,78 @@ func (q *Deque) Rotate(n int) {
 	}
 }
 
+// IndexFunc returns the index into the Deque of the first item satisfying
+// f(item), or -1 if none do.
+func (q *Deque) Index(f func(interface{}) bool) int {
+	if q.Len() > 0 {
+		modBits := len(q.buf) - 1
+		for i := 0; i < q.count; i++ {
+			if f(q.buf[(q.head+i)&modBits]) {
+				return i
+			}
+		}
+	}
+	return -1
+}
+
+// Insert is used to insert an element into the middle of the queue, before the
+// element at the specified index.  Insert(0,e) is the same as PushFront(e) and
+// Insert(Len(),e) is the same as PushBack(e).  Complexity is constant plus
+// linear in the lesser of the distances between `at` and either of the ends of
+// the queue.  Accepts only non-negative index values, and panics if index is
+// out of range.
+func (q *Deque) Insert(at int, item interface{}) {
+	if at < 0 || at > q.count {
+		panic("deque: Insert() called with index out of range")
+	}
+	if at*2 < q.count {
+		q.PushFront(item)
+		front := q.head
+		for i := 0; i < at; i++ {
+			next := q.next(front)
+			q.buf[front], q.buf[next] = q.buf[next], q.buf[front]
+			front = next
+		}
+		return
+	}
+	swaps := q.count - at
+	q.PushBack(item)
+	back := q.prev(q.tail)
+	for i := 0; i < swaps; i++ {
+		prev := q.prev(back)
+		q.buf[back], q.buf[prev] = q.buf[prev], q.buf[back]
+		back = prev
+	}
+}
+
+// Remove removes and returns an element from the middle of the queue, at the
+// specified index.  Remove(0) is the same as PopFront() and Remove(Len()-1) is
+// the same as PopBack().  Complexity is constant plus linear in the lesser of
+// the distances between i and either of the ends of the queue.  Accepts only
+// non-negative index values, and panics if index is out of range.
+func (q *Deque) Remove(at int) interface{} {
+	if at < 0 || at >= q.Len() {
+		panic("deque: Remove() called with index out of range")
+	}
+
+	rm := (q.head + at) & (len(q.buf) - 1)
+	if at*2 < q.count {
+		for i := 0; i < at; i++ {
+			prev := q.prev(rm)
+			q.buf[prev], q.buf[rm] = q.buf[rm], q.buf[prev]
```

Comment:
The statement here performs a swap.  In go, you can do the swap without explicitly creating a temporary.

The `for` loop moves the item to remove from inside the queue to the front of the queue, where it will finally be removed using `PopFront()`.